keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,280 @@
+
+### Revoke Bucket Access (needs to handle stratic and dynamic brownfield!)
+
+Reminder: BucketClass is ignored in browfield operations.
+
+1. The user deletes the OBC, which blocks until the finalizer is removed.
+1. The central controller detects the OBC change and updates OB’s Phase.
+1. The Sidecar detects the OB update and calls the `Revoke()` gRPC method.
+1. The Driver revokes access to the bucket and responds with no errors.
+1. The Sidecar deletes the OB’s secret.
+1. The central controller deletes the secret’s copy in the app’s namespace.
+1. The central controller removes the finalizers from the OBC, OB, and Secret.
+1. The OBC and its secret are garbage collected and the OB remains.
+
+##  Custom Resource Definitions
+
+#### ObjectBucketClaim (OBC)
+
+A user facing API object representing a request for a bucket, or access to an existing bucket. OBCs are created by users in their namespaces. Once the request is fulfilled, the OBC is “bound” to an Object Bucket (OB). The binding is used to mark the request as fulfilled and prevent further binds to the OB.
+
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: ObjectBucketClaim
+metadata:
+  name: [1]
+  namespace:
+  labels:
+    “cosi.io/driver”: [2]
+  finalizers:
+  - cosi.io/finalizer [3]
+spec:
+  bucketName: [4]
+  generateBucketName: [5]
+  bucketClassName: [6]
+  objectBucketName: [7]
+  secretName: [8]
+  driverSecretName: [9]
+status:
+  phase:
+  conditions: []ObjectBucketClaimCondition
+```
+1. `name`: the metadata name of the OBC; however this name can be generated and thus cannot be relied on to predictably name other related resources, eg. secrets. 
+1. `labels`: central controller adds the label to its managed resources for easy GET ops.  Value is the driver name returned by GetDriverInfo() rpc*.
+1. `finalizers`: central controller adds the finalizer to defer OBC deletion until backend deletion ops succeed.
+1. `bucketName`: Desired name of the bucket to be created**.
+1. `generateBucketName`: Prefix to a randomly generated name. Mutually exclusive with `bucketName`**.
+1. `bucketClassName`: Name of the target BucketClass**.
+1. `objectBucketName`: Name of a bound OB.
+   - Injected by the central controller during greenfield ops.
+   - Defined by the OBC author for static and dynamic brownfield ops.
+1. `secretName`: Desired name of the app's secret for greenfield. Fails if secret exists. Defined here so that app deployment (where the secret name is required) is independent of OBC creation.
+1. `driverSecretName`: Name of the driver's secret with the namespace assumed to be "cosi-system".
+   - Injected by the central controller during greenfield and dynamic brownfield ops.
+   - Defined by the OBC author for static brownfield ops.
+
+\* Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
```

Comment:
How do you deal with conflicts?

Reply:
Reasons for conflicts: 1) two drivers return the same name, 2) we replace offending driver name chars with "-" and create our own conflict.
The first case could be intentional. The second case would appear to be a bug. Is there a better "solution" than documenting legal driver names?

**Note**: if we go back to proposing that drivers are isolated by ns, then the value should include the ns but conflicts could still occur.

Reply:
This could be handled by registering the drivers as they're installed and on COSI controller start.  

Registration could be accomplished in a couple ways e.g.:

Using a ProvisionerRegistration object to abstract the provisioner.  On ADD events, COSI verifies that the driver name isn't already taken.  

Or using a reverse facing gRPC interface implemented by the COSI controller.  Sidecars carry the client for this interface and on start, get the driver data and rpc it to the COSI controller, which would be at a known network endpoint. 

In either case the sidecar needs to be given a signal that it may begin processing OBs.  This could be a flag on in the ProvisionerRegistration.status or a return value of the rpc.

I realize this is more complicated than CSI's registration.  COSI doesn't have access to the kubelet and isn't bound by node affinity, meaning unix sockets aren't useful as a solution.